### ISSUE TITLE
Grails 4.0.1 data binding not working ?

### ISSUE THREAD

**Issue by abcfy2** [grails/grails-core#11423]
Thanks for reporting an issue for Grails framework, please review the task list below before submitting the 
issue. Your issue report will be closed if the issue is incomplete and the below tasks not completed.

NOTE: If you are unsure about something and the issue is more of a question a better place to ask questions is on Stack Overflow (http://stackoverflow.com/tags/grails) or Slack (http://slack-signup.grails.org). DO NOT use the issue tracker to ask questions.

### Task List

- [x] Steps to reproduce provided
- [ ] Stacktrace (if present) provided
- [x] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

### Steps to Reproduce

Grails doc: https://docs.grails.org/4.0.1/guide/single.html#_date_formats_for_data_binding:

```
grails-app/conf/application.groovy
grails.databinding.dateFormats = ['MMddyyyy', 'yyyy-MM-dd HH:mm:ss.S', "yyyy-MM-dd'T'hh:mm:ss'Z'"]
```

But I find the code: 

https://github.com/grails/grails-core/blob/b3c66011fada9528436d946a32dabcb3e422644b/grails-plugin-databinding/src/main/groovy/org/grails/plugins/databinding/DataBindingConfigurationProperties.java#L30-L37

All the `grails.databinding` configs are steadfast by default value and doesn't change anymore.

But the deprecated `AbstractDataBindingGrailsPlugin` has a `doWithSpring` closure to modify the default value:

https://github.com/grails/grails-core/blob/b3c66011fada9528436d946a32dabcb3e422644b/grails-plugin-databinding/src/main/groovy/org/grails/plugins/databinding/AbstractDataBindingGrailsPlugin.groovy#L57-L75

Then I try to add `grails.databinding.dateFormats = ["yyyy-MM-dd HH:mm:ss"]` to `application.groovy` and testing in IDEA Intellij, and I find the `formatStrings` in `Jsr310ConvertersConfiguration.groovy` does not change to my setting.

### Expected Behaviour

`grails.databing` should be working.

### Actual Behaviour

`grails.databing` seems broken.

### Environment Information

- **Operating System**: Windows 10
- **Grails Version:** 4.0.1
- **JDK Version:** 1.8
- **Container Version (If Applicable):** TODO

**Comment by graemerocher** [comment]
Probably as a workaround you can either add `micronaut-runtime-groovy` to your classpath or specify the values in `application.yml`

**Comment by abcfy2** [comment]
After add `compile "io.micronaut:micronaut-runtime-groovy:1.1.1"` to `build.gradle`, will cause this error:

```
java.lang.IllegalStateException: Failed to load ApplicationContext
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:125)
	at org.springframework.test.context.support.DefaultTestContext.getApplicationContext(DefaultTestContext.java:108)
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.injectDependencies(DependencyInjectionTestExecutionListener.java:118)
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.prepareTestInstance(DependencyInjectionTestExecutionListener.java:83)
	at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:246)
	at org.spockframework.spring.SpringTestContextManager.prepareTestInstance(SpringTestContextManager.java:56)
	at org.spockframework.spring.SpringInterceptor.interceptInitializerMethod(SpringInterceptor.java:43)
	at org.spockframework.runtime.extension.AbstractMethodInterceptor.intercept(AbstractMethodInterceptor.java:24)
	at org.spockframework.runtime.extension.MethodInvocation.proceed(MethodInvocation.java:97)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:118)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:175)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:157)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:404)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
	at java.lang.Thread.run(Thread.java:748)
Caused by: io.micronaut.context.exceptions.ConfigurationException: Exception occurred reading configuration [application]: startup failed:
Script1.groovy: 2: [Static type checking] - Cannot find matching method Script1#id(java.util.LinkedHashMap <java.lang.String, java.lang.String>). Please check if the declared type is correct and if the method exists.
 @ line 2, column 5.
       id generator: 'identity'
       ^

1 error

	at io.micronaut.context.env.groovy.GroovyPropertySourceLoader.processInput(GroovyPropertySourceLoader.java:54)
	at io.micronaut.context.env.AbstractPropertySourceLoader.read(AbstractPropertySourceLoader.java:110)
	at io.micronaut.context.env.AbstractPropertySourceLoader.loadProperties(AbstractPropertySourceLoader.java:95)
	at io.micronaut.context.env.AbstractPropertySourceLoader.load(AbstractPropertySourceLoader.java:65)
	at io.micronaut.context.env.DefaultEnvironment.loadPropertySourceFromLoader(DefaultEnvironment.java:582)
	at io.micronaut.context.env.DefaultEnvironment.readPropertySourceList(DefaultEnvironment.java:523)
	at io.micronaut.context.DefaultApplicationContext$RuntimeConfiguredEnvironment.readPropertySourceList(DefaultApplicationContext.java:648)
	at io.micronaut.context.env.DefaultEnvironment.readPropertySources(DefaultEnvironment.java:448)
	at io.micronaut.context.env.DefaultEnvironment.start(DefaultEnvironment.java:311)
	at io.micronaut.context.env.DefaultEnvironment.start(DefaultEnvironment.java:63)
	at io.micronaut.context.DefaultApplicationContext.startEnvironment(DefaultApplicationContext.java:228)
	at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:186)
	at grails.boot.GrailsApp.createApplicationContext(GrailsApp.groovy:155)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:308)
	at grails.boot.GrailsApp.run(GrailsApp.groovy:97)
	at org.springframework.boot.test.context.SpringBootContextLoader.loadContext(SpringBootContextLoader.java:120)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:99)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:117)
	... 29 more
Caused by: org.codehaus.groovy.control.MultipleCompilationErrorsException: startup failed:
Script1.groovy: 2: [Static type checking] - Cannot find matching method Script1#id(java.util.LinkedHashMap <java.lang.String, java.lang.String>). Please check if the declared type is correct and if the method exists.
 @ line 2, column 5.
       id generator: 'identity'
       ^

1 error

	at org.codehaus.groovy.control.ErrorCollector.failIfErrors(ErrorCollector.java:311)
	at org.codehaus.groovy.control.CompilationUnit.applyToPrimaryClassNodes(CompilationUnit.java:1102)
	at org.codehaus.groovy.control.CompilationUnit.doPhaseOperation(CompilationUnit.java:645)
	at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:623)
	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:600)
	at groovy.lang.GroovyClassLoader.doParseClass(GroovyClassLoader.java:390)
	at groovy.lang.GroovyClassLoader.access$300(GroovyClassLoader.java:89)
	at groovy.lang.GroovyClassLoader$5.provide(GroovyClassLoader.java:330)
	at groovy.lang.GroovyClassLoader$5.provide(GroovyClassLoader.java:327)
	at groovy.lang.GroovyClassLoader.parseClass(GroovyClassLoader.java:325)
	at groovy.lang.GroovyShell.parseClass(GroovyShell.java:540)
	at groovy.lang.GroovyShell.parse(GroovyShell.java:552)
	at groovy.lang.GroovyShell.parse(GroovyShell.java:532)
	at groovy.lang.GroovyShell.parse(GroovyShell.java:597)
	at io.micronaut.context.env.groovy.ConfigurationEvaluator.evaluate(ConfigurationEvaluator.groovy:73)
	at io.micronaut.context.env.groovy.ConfigurationEvaluator.evaluate(ConfigurationEvaluator.groovy:47)
	at io.micronaut.context.env.groovy.GroovyPropertySourceLoader.processInput(GroovyPropertySourceLoader.java:49)
	... 46 more
```

application.groovy:

```groovy
grails.gorm.default.mapping = {
    id generator: 'identity'
}
```

Move this to `runtime.groovy`, compile passed, but seems `grails.databing` still not working.

And I also try to add config to `application.yml`, still not working.

Here is my project to reproduce this issue: https://github.com/abcfy2/grails-rest-seed

Steps:

- create postgresql role and database:

    ```sql
    CREATE ROLE earth_admin LOGIN PASSWORD 'admin';
    CREATE DATABASE earch_test OWNER earth_admin;
    ```
- run integration test:

    ```bash
    ./grailsw test-app -integration
    ```

**Comment by ZacharyKlein** [comment]
@graemerocher @jeffbrown I have created a new sample project that I think more concisely demonstrates the issue. In my testing, adding the `micronaut-runtime-groovy` dependency does resolve the problem with databinding of custom date formats.    

However (as I note in my sample), the binding behavior between the running app and a test using the `ControllerUnitTest` trait is not consistent (the databinding works in the test - with or without the extra dependency - whereas the application fails to bind correctly without the dependency).

I was able to reliable demonstrate the issue (and fix/workaround via the extra dependency) using a `@Integration` test, which I have included in the sample project.

https://github.com/ZacharyKlein/grails_issue_11423_databinding
